### PR TITLE
Donations Contract Optimisation

### PR DESCRIPTION
**Pull request by realdiganta**
1. Just a few minor gas optimisations
2. Better packing of the Metadata struct to save on gas
3. burnBatch method to burn multiple nft ids together
4. Tests for the burnBatch method

### CODE REVIEW

File: contracts/Donations.sol [hunk 1]
```diff
@@ -91,16 +91,13 @@ contract Donations is ERC721, AccessControl {
      * @param _to Address of the charity.
      */
     function donate(
-        uint256 _destinationId,
+        uint128 _destinationId,
```
Comment (gabrielpoca) [inline contracts/Donations.sol:94]: Why this change?

Comment (realdiganta) [reply contracts/Donations.sol:94]: If you check the Metadata struct, then you will find that i have change the destinationId type there to uint128 from uint256. And I also moved the "uint64 expiry" struct to 2nd position from 4th position. What this will do this pack the destinationId & the expiry variables in the same 32 bytes slot instead of using 2 slots for them.  Due to this change I also had to change the transferableAmounts mapping to use uint128 for the destinationId. Thus, I changed it to uint128 in the donate method too. Else if it was uint256 here, then we would have to cast it to uint128 again to use in the transferableAmounts  update on line 101. ( I am not totally sure about the last part, maybe if we kept it at uint256 then solidity would internally cast it to uint128 on line 101, but not sure if that would use more gas or not)

Comment (gabrielpoca) [reply contracts/Donations.sol:94]: sorry, I wasn't asking about this line specifically, but I understand that you did it to be able to pack it with the other variable. My concern is that I'm not sure 128 is enough, [their API](https://the-giving-block.gitbook.io/public-api-documentation/endpoints-documentation) doesn't specify the size of the ID

Comment (CrisBRM) [reply contracts/Donations.sol:94]: let's ask them if there's a cap :]

Comment (realdiganta) [reply contracts/Donations.sol:94]: I guess they wont make ids larger than 340282366920938463463374607431768211456 (2**128). But not sure, definitely a good idea to be double sure by asking them.

Comment (CrisBRM) [reply contracts/Donations.sol:94]: Alex from TGB said he doesn't think so (and if he comes back and rectifies it we can fix it anyway)

Comment (realdiganta) [reply contracts/Donations.sol:94]: so what do we do ? uint256, uint128 or we can also try uint192? Since uint192 will still total up to 256 when packed with uint64.

Comment (CrisBRM) [reply contracts/Donations.sol:94]: either one but it must add up to 256 or it will use more gas

File: contracts/Donations.sol [hunk 2]
```diff
@@ -175,6 +175,33 @@ contract Donations is ERC721, AccessControl {
...
+     * Expected to be called by the backend to burn a bunch of NFTs when they expire
+     *
+     * @param _ids IDs of the NFTs.
+     */
+    function burnBatch(uint256[] calldata _ids) external {
+        uint256 _id;
+        Metadata storage data;
+        bool expired;
+
+        for (uint256 i = 0; i < _ids.length; ++i) {
```
Comment (CrisBRM) [inline contracts/Donations.sol:191]: cache length outside of loop pls, something like

```
uint256 idsLen = _ids.length
for (uint256 i = 0; i < idsLen; ++i)
```

File: contracts/Donations.sol (hunk 1 above)
Comment (CrisBRM) [reply contracts/Donations.sol:94]: You can keep it as is. It's been confirmed 👍

**Comment by naps62** [comment]
Merged #121 instead
